src/gwt/src/org/rstudio/core/client/VirtualConsole.java

```diff
@@ -203,99 +126,331 @@ else if (string.length() > maxLength)
    
    public int getLength()
    {
-      return o.length();
+      return output_.length();
+   }
+   
+   public static String consolify(String text)
+   {
+      VirtualConsole console = new VirtualConsole();
+      console.submit(text);
+      return console.toString();
+   }
+   
+   public Element getParent()
+   {
+      return parent_;
    }
    
-   public void submitAndRender(String data, String clazz, Element parent)
+   /**
+    * Appends text to the end of the virtual console.
+    * 
+    * @param text The text to append
+    * @param clazz The content to append to the text.
+    */
+   private void appendText(String text, String clazz)
    {
-      if (!submit(data, clazz))
+      Entry <Integer, ClassRange> last = class_.lastEntry();
+      ClassRange range = last.getValue();
+      if (range.clazz == clazz)
       {
-         // output isn't append-only; redraw the whole thing
-         // (note that even this isn't technically necessary but control 
-         // characters are relatively infrequent and additional bookkeeping
-         // would be required to determine the invalidated range when 
-         // control characters are used)
-         redraw(parent);
+         // just append to the existing output stream
+         range.appendRight(text, 0);
       }
       else
       {
-         // consolify just the data to be rendered
-         emitRange(consolify(data), clazz, parent);
+         // create a new output range with this class
+         final ClassRange newRange = new ClassRange(cursor_, clazz, text);
+         parent_.appendChild(newRange.element);
```

Comment:
Good question; it does not -- we only call this function when there is a `parent_`. https://github.com/rstudio/rstudio/pull/1035/files#diff-5c5c61a14b365d8fe55a849f020f72ebR318

(some existing `parent_` tests may be superfluous)

This distinguishes the VirtualConsole's "live update" mode (where output is rendered to the DOM as it is received) from the "consolify" case (where a temporary console is constructed, outside the DOM, just to process escape sequences in string).